modules/alloy/src/main/ts/ephox/alloy/positioning/layout/PinnedLayout.ts

```diff
@@ -0,0 +1,42 @@
+import { nu as NuSpotInfo } from '../view/SpotInfo';
+import { Bubble } from './Bubble';
+import * as Direction from './Direction';
+import { boundsRestriction } from './LayoutBounds';
+import { northPinned, southPinned } from './LayoutLabels';
+import { AnchorBox, AnchorElement, AnchorLayout } from './LayoutTypes';
+
+const pinAtTop: AnchorLayout = (
+  anchor: AnchorBox,
+  element: AnchorElement,
+  bubbles: Bubble
+) => {
+  return NuSpotInfo(
+    // cap the bounds.
+    anchor.x + anchor.width / 2 - element.width / 2,
```

Comment:
Would it be worth trying to share this logic with `Layout.ts`? There's a bunch of little utility functions that handle exactly this sort of thing in there, and `middleX` reads a lot easier than `anchor.x + anchor.width / 2 - element.width / 2` for trying to figure out where the spot is going to be placed.

Reply:
Layout and LayoutInside, the previously existing versions, have their own separate implementations as-is. I'll make it to separate functions in PinnedLayout though, for ease of readability.